### ISSUE TITLE
Heart Rate Smoothing

### ISSUE THREAD

**Issue by wbrenna** [jonasoreland/runnerup#279]
Is heart rate smoothing anything anyone else wants? I've tested out an [exponential smoothing](https://en.wikipedia.org/wiki/Exponential_smoothing) with alpha=0.35 and it solves the problems with my bluetooth HRM (namely, when the connection is weak the reading spikes to ~220 or drops to zero). I could fairly easily set up a preference for alpha (default 1 = no smoothing) if this is at all desired.

**Comment by jonasoreland** [comment]
i think it might be interesting...
i already have a bunch of smoothing functions for pace.

**Comment by wbrenna** [comment]
Cool, I'll test it as I add the preference slowly, to avoid any bugs.

The smoothing implementation is quite small - you can take a look here: wbrenna/runnerup@5864ea030868ca7ea63e97175071493788ad7e5e.

**Comment by jonasoreland** [comment]
don't you think smoothing should be in presentation layer ? not in
capturing...

or where is your HR problem, when looking at graph or somehwere else ?

/Jonas

**Comment by wbrenna** [comment]
I use third party tools to view the HR data, which don't have any smoothing features. Because the HR spikes for my device are spurious rather than just unappealing, my preference was to eliminate them from the capture entirely. I can understand if you'd rather not give the option to tamper with the raw data, though.

If it were implemented in the presentation layer, would it still transmit the raw HR data when uploaded to third-party sites?

**Comment by jonasoreland** [comment]
re presentation layer: not automatically, no

so...we can but smoothing in any of the following places
- capturing (affects everything)
- presentation (only affects graph in RunnerUp)
- upload (affects all sites being uploaded to)

i'm not sure what I think...
capturing might not be that bad...if these devices in general are shaky.
but then we really should choose a good alpha value since very few will
modify a default value.

how about this ??
1) add a configurable smoothing function to presentation layer (like it's
done for pace)
then we could experiment some to choose a good alpha value

2) then we add it to capturing (still with alpha being configrable)

/Jonas

**Comment by nebmo** [comment]
I agree about wanting it at least for presentation and upload (or capturing). I usually look at HR values on connected sites and not in runnerup

**Comment by wbrenna** [comment]
My thoughts were to make the default alpha = 1. This means no smoothing, all data is captured as original. Then, if someone wants to smooth, they can manually change alpha to whatever they'd like in the preferences.

I'm happy to test it out in the presentation layer though.
